File: lib/wasm-micro-runtime-fast-jit-06-29-2022/assembly-script/wamr_app_lib/timer.ts
/*
 * Copyright (C) 2019 Intel Corporation.  All rights reserved.
 * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 */

@external("env", "wasm_create_timer")
declare function wasm_create_timer(a: i32, b: bool, c: bool): i32;

@external("env", "wasm_timer_cancel")
declare function wasm_timer_cancel(a: i32): void;

@external("env", "wasm_timer_restart")
declare function wasm_timer_restart(a: i32, b: i32): void;

@external("env", "wasm_get_sys_tick_ms")
declare function wasm_get_sys_tick_ms(): i32;

export var timer_list = new Array<user_timer>();

export class user_timer {
    timer_id: i32 = 0;
    timeout: i32;
    period: bool = false;
    cb: () => void;

    constructor(cb: () => void, timeout: i32, period: bool) {
        this.cb = cb;
        this.timeout = timeout;
        this.period = period
        this.timer_id = timer_create(this.timeout, this.period, true);
    }
}

export function timer_create(a: i32, b: bool, c: bool): i32 {
    return wasm_create_timer(a, b, c);
}

export function setTimeout(cb: () => void, timeout: i32): user_timer {
    var timer = new user_timer(cb, timeout, false);
    timer_list.push(timer);

    return timer;
}

export function setInterval(cb: () => void, timeout: i32): user_timer {
    var timer = new user_timer(cb, timeout, true);
    timer_list.push(timer);

    return timer;
}

export function timer_cancel(timer: user_timer): void {
    wasm_timer_cancel(timer.timer_id);

    var i = 0;
    for (i = 0; i < timer_list.length; i++) {
        if (timer_list[i].timer_id == timer.timer_id)
            break;
    }

    timer_list.splice(i, 1);
}

export function timer_restart(timer: user_timer, interval: number): void {
    wasm_timer_restart(timer.timer_id, i32(interval));
}

export function now(): i32 {
    return wasm_get_sys_tick_ms();
}

// This export function need to be copied to the top application file
//
export function on_timer_callback(on_timer_id: i32): void {
    for (let i = 0; i < timer_list.length; i++) {
        if (timer_list[i].timer_id == on_timer_id) {
            timer_list[i].cb();
        }
    }
}

File: lib/wasm-micro-runtime-WAMR-1.3.3/core/app-framework/base/app/timer.c
/*
 * Copyright (C) 2019 Intel Corporation.  All rights reserved.
 * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 */

#include <stdlib.h>
#include <string.h>

#include "wa-inc/timer_wasm_app.h"
#include "timer_api.h"

#if 1
#include <stdio.h>
#else
#define printf (...)
#endif

struct user_timer {
    struct user_timer *next;
    int timer_id;
    void (*user_timer_callback)(user_timer_t);
};

struct user_timer *g_timers = NULL;

user_timer_t
api_timer_create(int interval, bool is_period, bool auto_start,
                 on_user_timer_update_f on_timer_update)
{

    int timer_id = wasm_create_timer(interval, is_period, auto_start);

    // TODO
    struct user_timer *timer =
        (struct user_timer *)malloc(sizeof(struct user_timer));
    if (timer == NULL) {
        // TODO: remove the timer_id
        printf("### api_timer_create malloc faild!!! \n");
        return NULL;
    }

    memset(timer, 0, sizeof(*timer));
    timer->timer_id = timer_id;
    timer->user_timer_callback = on_timer_update;

    if (g_timers == NULL)
        g_timers = timer;
    else {
        timer->next = g_timers;
        g_timers = timer;
    }

    return timer;
}

void
api_timer_cancel(user_timer_t timer)
{
    user_timer_t t = g_timers, prev = NULL;

    wasm_timer_cancel(timer->timer_id);

    while (t) {
        if (t == timer) {
            if (prev == NULL) {
                g_timers = t->next;
                free(t);
            }
            else {
                prev->next = t->next;
                free(t);
            }
            return;
        }
        else {
            prev = t;
            t = t->next;
        }
    }
}

void
api_timer_restart(user_timer_t timer, int interval)
{
    wasm_timer_restart(timer->timer_id, interval);
}

void
on_timer_callback(int timer_id)
{
    struct user_timer *t = g_timers;

    while (t) {
        if (t->timer_id == timer_id) {
            t->user_timer_callback(t);
            break;
        }
        t = t->next;
    }
}
